feat: k/n memory manager updates (#3861)
JetBrains/kotlin-web-site: docs/topics/native/native-memory-manager.md

## Garbage collector

The exact algorithm of GC is constantly evolving. Currently, it is the Stop-the-World Mark and Concurrent Sweep
collector that does not separate the heap into generations.

The GC uses a full parallel mark that combines paused mutators, the GC thread, and optional marker threads to process
the mark queue. By default, paused mutators and at least one GC thread participate in the marking process.
You can disable the full parallel mark with `-Xbinary=gcMarkSingleThreaded=true` compilation option.
However, it may increase the pause time of the garbage collector.

When the marking phase is completed, GC processes weak references and nullifies reference points to an unmarked object.

> To decrease the GC pause time, you can enable the concurrent processing of weak references. To do that, use

Reviewer: ```suggestion
To decrease the GC pause time, you can enable the concurrent processing of weak references by using
```

## Memory consumption

Kotlin/Native uses its own [memory allocator](https://github.com/JetBrains/kotlin/blob/master/kotlin-native/runtime/src/alloc/custom/README.md).
It divides system memory into pages, allowing independent sweeping in consecutive order. Each allocation becomes a memory
block within a page, and the page keeps track of block sizes. Different page types are optimized for various allocation
sizes. The consecutive arrangement of memory blocks ensures efficient iteration through all allocated blocks.

When a thread allocates memory, it searches for a suitable page based on the allocation size. Threads maintain a set of
pages for different size categories. Typically, the current page for a given size can accommodate the allocation.
If not, the thread requests a different page from the shared allocation space. This page may already be available,
require sweeping, or have to be created first.

The Kotlin/Native memory allocator comes with protection against sudden spikes in memory allocations. It prevents

> situations when the mutator starts to allocate a lot of garbage quickly, and the GC thread cannot keep up with it,

Reviewer: ```suggestion
situations where the mutator starts to allocate a lot of garbage quickly, and the GC thread cannot keep up with it,
```

try updating Kotlin to the latest version. We're constantly improving the memory manager, so even a simple compiler
update might improve memory consumption.

If you experience high memory consumption anyway, a few options are available:

> * Switch to a different memory allocator. To do that, use the following compilation options in your Gradle build script:

Reviewer: ```suggestion
* Switch to a different memory allocator by using one of the following compilation options in your Gradle build script:
```
